<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>10X Single Cell RNA Sequencing &mdash; Giotto 1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx-toolbox-code.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/sphinx-toolbox.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/tabs.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx_toolbox_installation.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script src="../../_static/sphinx_toolbox_installation.js"></script>
        <script src="../../_static/design-tabs.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html">
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contribute.html">Contribute</a></li>
<li class="toctree-l1"><a class="reference external" href="http://spatial.rc.fas.harvard.edu/giotto-viewer/">Giotto Viewer [work-in-progress]</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html">Datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../trygiotto.html">Try Giotto</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Issues</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Common Errors and Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../github_issues.html">Report a Bug</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>10X Single Cell RNA Sequencing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/subsections/datasets/singlecell_prostate_standard.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="x-single-cell-rna-sequencing">
<h1>10X Single Cell RNA Sequencing<a class="headerlink" href="#x-single-cell-rna-sequencing" title="Permalink to this headline"></a></h1>
<section id="set-up-giotto-environment">
<h2>Set up Giotto Environment<a class="headerlink" href="#set-up-giotto-environment" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>

<span class="c1"># 1. set working directory</span>
<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/result&#39;</span>

<span class="c1"># 2. set giotto python path</span>
<span class="c1"># set python path to your preferred python version path</span>
<span class="c1"># set python path to conda env/bin/ directory if manually installed Giotto python dependencies by conda</span>
<span class="c1"># python_path = &#39;/path_to_conda/.conda/envs/giotto/bin/python&#39;</span>
<span class="c1"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span>
<span class="n">python_path</span> <span class="o">=</span> <span class="n">NULL</span>
<span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">python_path</span><span class="p">))</span> <span class="p">{</span>
<span class="n">installGiottoEnvironment</span><span class="p">()</span>
<span class="p">}</span>

<span class="c1"># 3. create giotto instructions</span>
<span class="n">instrs</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span>
                                <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                                <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                                <span class="n">python_path</span> <span class="o">=</span> <span class="n">python_path</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="dataset-explanation">
<h2>Dataset Explanation<a class="headerlink" href="#dataset-explanation" title="Permalink to this headline"></a></h2>
<p><a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/33032611/">Ma et al.</a> Processed 10X Single Cell RNAseq from two prostate cancer patients. The raw dataset can be found <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157703">here</a></p>
</section>
<section id="part-1-create-giotto-object-from-10x-dataset">
<h2>Part 1: Create Giotto object from 10X dataset<a class="headerlink" href="#part-1-create-giotto-object-from-10x-dataset" title="Permalink to this headline"></a></h2>
<p>Note that you will need an input directory for <cite>barcodes.tsv(.gz)</cite> <cite>features.tsv(.gz)</cite> <cite>matrix.mtx(.gz)</cite></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">giotto_SC</span><span class="o">&lt;-</span><span class="n">createGiottoObject</span><span class="p">(</span><span class="n">expression</span> <span class="o">=</span> <span class="n">get10Xmatrix</span><span class="p">(</span><span class="s2">&quot;/path/to/filtered_feature_bc_matrix&quot;</span><span class="p">,</span>
              <span class="n">gene_column_index</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">remove_zero_rows</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">),</span>
              <span class="n">instructions</span> <span class="o">=</span> <span class="n">instrs</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="part-2-process-giotto-object">
<h2>Part 2: Process Giotto Object<a class="headerlink" href="#part-2-process-giotto-object" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    giotto_SC&lt;-filterGiotto(gobject = giotto_SC,
expression_threshold = 1,
feat_det_in_min_cells = 50,
min_det_feats_per_cell = 500,
expression_values = c(&#39;raw&#39;),
verbose = T)

    ## normalize
    giotto_SC &lt;- normalizeGiotto(gobject = giotto_SC, scalefactor = 6000)

    ## add mitochondria gene percentage and filter giotto object by percent mito
    library(rtracklayer)
    gtf &lt;- import(&quot;Homo_sapiens.GRCh38.105.gtf.gz&quot;)
    gtf &lt;- gtf[gtf$gene_name!=&quot;&quot; &amp; !is.na(gtf$gene_name)]
    mito &lt;- gtf$gene_name[as.character(seqnames(gtf)) %in% &quot;MT&quot;]
    mito&lt;-unique(mito)

    giotto_SC&lt;-addFeatsPerc(
    giotto_SC,
    feats = mito,
    vector_name = &#39;perc_mito&#39;
    )

    giotto_SC&lt;-subsetGiotto(giotto_SC,
    cell_ids = pDataDT(giotto_SC)[which(pDataDT(giotto_SC)$perc_mito &lt; 15),]$cell_ID)


    ## add gene &amp; cell statistics
    giotto_SC &lt;- addStatistics(gobject = giotto_SC, expression_values = &#39;raw&#39;)
</pre></div>
</div>
</section>
<section id="part-3-dimention-reduction">
<h2>Part 3: Dimention Reduction<a class="headerlink" href="#part-3-dimention-reduction" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## PCA ##</span>
<span class="n">giotto_SC</span> <span class="o">&lt;-</span> <span class="n">calculateHVF</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">)</span>
<span class="n">giotto_SC</span> <span class="o">&lt;-</span> <span class="n">runPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">scale_unit</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="n">screePlot</span><span class="p">(</span><span class="n">giotto_SC</span><span class="p">,</span> <span class="n">ncp</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_scree_plot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/3_scree_plot.png" src="../../_images/3_scree_plot.png" />
</section>
<section id="part-4-cluster">
<h2>Part 4: Cluster<a class="headerlink" href="#part-4-cluster" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## cluster and run UMAP ##</span>
<span class="c1"># sNN network (default)</span>
<span class="n">showGiottoDimRed</span><span class="p">(</span><span class="n">giotto_SC</span><span class="p">)</span>
<span class="n">giotto_SC</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span>
    <span class="n">dim_reduction_to_use</span> <span class="o">=</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">dim_reduction_name</span> <span class="o">=</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span>
    <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>

<span class="c1"># UMAP</span>
<span class="n">giotto_SC</span> <span class="o">=</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">giotto_SC</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Leiden clustering</span>
<span class="n">giotto_SC</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>


<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s2">&quot;4_Cluster&quot;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/4_Cluster.png" src="../../_images/4_Cluster.png" />
</section>
<section id="part-5-differential-expression">
<h2>Part 5: Differential Expression<a class="headerlink" href="#part-5-differential-expression" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">markers_scran</span> <span class="o">=</span> <span class="n">findMarkers_one_vs_all</span><span class="p">(</span><span class="n">gobject</span><span class="o">=</span><span class="n">giotto_SC</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;scran&quot;</span><span class="p">,</span>
                   <span class="n">expression_values</span><span class="o">=</span><span class="s2">&quot;normalized&quot;</span><span class="p">,</span> <span class="n">cluster_column</span><span class="o">=</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">min_feats</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">markergenes_scran</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">markers_scran</span><span class="p">[,</span> <span class="n">head</span><span class="p">(</span><span class="o">.</span><span class="n">SD</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">][[</span><span class="s2">&quot;feats&quot;</span><span class="p">]])</span>

<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">giotto_SC</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s2">&quot;normalized&quot;</span><span class="p">,</span> <span class="n">metadata_cols</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
                    <span class="n">selected_feats</span> <span class="o">=</span> <span class="n">markergenes_scran</span><span class="p">,</span>
                    <span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">show_values</span> <span class="o">=</span> <span class="s1">&#39;zscores_rescaled&#39;</span><span class="p">,</span>
                    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_a_metaheatmap&#39;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/5_a_metaheatmap.png" src="../../_images/5_a_metaheatmap.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>topgenes_scran = markers_scran[, head(.SD, 1), by = &#39;cluster&#39;]$feats
# violinplot
violinPlot(giotto_SC, feats = unique(topgenes_scran), cluster_column = &#39;leiden_clus&#39;,
        strip_text = 10, strip_position = &#39;right&#39;,
        save_param = list(save_name = &#39;5_b_violinplot_scran&#39;, base_width = 5))
</pre></div>
</div>
<img alt="../../_images/5_b_violinplot_scran.png" src="../../_images/5_b_violinplot_scran.png" />
</section>
<section id="part-6-featureplot">
<h2>Part 6: FeaturePlot<a class="headerlink" href="#part-6-featureplot" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot known marker genes across different cell types. EPCAM for epithelial cells,</span>
<span class="c1"># DPP4(CD26) for Epithelial luminal cells, PECAM1(CD31) for Endothelial cells and CD3D for T cells</span>
<span class="n">dimFeatPlot2D</span><span class="p">(</span><span class="n">giotto_SC</span><span class="p">,</span> <span class="n">feats</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;EPCAM&quot;</span><span class="p">,</span><span class="s2">&quot;DPP4&quot;</span><span class="p">,</span><span class="s2">&quot;PECAM1&quot;</span><span class="p">,</span><span class="s2">&quot;CD3D&quot;</span><span class="p">),</span> <span class="n">cow_n_col</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s2">&quot;6_featureplot&quot;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/6_featureplot.png" src="../../_images/6_featureplot.png" />
</section>
<section id="part-7-cell-type-annotation">
<h2>Part 7: Cell type Annotation<a class="headerlink" href="#part-7-cell-type-annotation" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">prostate_labels</span><span class="o">&lt;-</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Endothelial cells&quot;</span><span class="p">,</span><span class="c1">#1</span>
       <span class="s2">&quot;T cells&quot;</span><span class="p">,</span><span class="c1">#2</span>
       <span class="s2">&quot;Epithelial_basal&quot;</span><span class="p">,</span><span class="c1">#3</span>
       <span class="s2">&quot;Epithelial_luminal&quot;</span><span class="p">,</span><span class="c1">#4</span>
       <span class="s2">&quot;Fibroblasts&quot;</span><span class="p">,</span><span class="c1">#5</span>
       <span class="s2">&quot;T cells&quot;</span><span class="p">,</span><span class="c1">#6</span>
       <span class="s2">&quot;Epithelial_luminal&quot;</span><span class="p">,</span><span class="c1">#7</span>
       <span class="s2">&quot;Smooth muscle cells&quot;</span><span class="p">,</span><span class="c1">#8</span>
       <span class="s2">&quot;Macrophage &amp; B cells&quot;</span><span class="p">,</span><span class="c1">#9</span>
       <span class="s2">&quot;Fibroblasts&quot;</span><span class="p">,</span><span class="c1">#10</span>
       <span class="s2">&quot;Mast cells&quot;</span><span class="p">,</span><span class="c1">#11</span>
       <span class="s2">&quot;Mesenchymal cells&quot;</span><span class="p">,</span><span class="c1">#12</span>
       <span class="s2">&quot;Neural Progenitor cells&quot;</span><span class="p">)</span><span class="c1">#13</span>
<span class="n">names</span><span class="p">(</span><span class="n">prostate_labels</span><span class="p">)</span><span class="o">&lt;-</span><span class="mi">1</span><span class="p">:</span><span class="mi">13</span>
<span class="n">giotto_SC</span><span class="o">&lt;-</span><span class="n">annotateGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span> <span class="n">annotation_vector</span> <span class="o">=</span> <span class="n">prostate_labels</span> <span class="p">,</span>
                        <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;prostate_labels&#39;</span><span class="p">)</span>
<span class="n">dimPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">giotto_SC</span><span class="p">,</span>     <span class="n">dim_reduction_name</span> <span class="o">=</span> <span class="s1">&#39;umap&#39;</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s2">&quot;prostate_labels&quot;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s2">&quot;7_Annotation&quot;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/7_Annotation.png" src="../../_images/7_Annotation.png" />
</section>
<section id="part-8-subset-and-recluster">
<h2>Part 8: Subset and Recluster<a class="headerlink" href="#part-8-subset-and-recluster" title="Permalink to this headline"></a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>      Subset_giotto_T&lt;-subsetGiotto(giotto_SC,
cell_ids = pDataDT(giotto_SC)[which(pDataDT(giotto_SC)$prostate_labels == &quot;T cells&quot;),]$cell_ID)
      ## PCA

      Subset_giotto_T &lt;- calculateHVF(gobject = Subset_giotto_T)
      Subset_giotto_T &lt;- runPCA(gobject = Subset_giotto_T, center = TRUE, scale_unit = TRUE)
      screePlot(Subset_giotto_T, ncp = 20, save_param = list(save_name = &#39;8a_scree_plot&#39;))
</pre></div>
</div>
<img alt="../../_images/8a_scree_plot.png" src="../../_images/8a_scree_plot.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="n">Subset_giotto_T</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">Subset_giotto_T</span><span class="p">,</span>
<span class="n">dim_reduction_to_use</span> <span class="o">=</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span> <span class="n">dim_reduction_name</span> <span class="o">=</span> <span class="s1">&#39;pca&#39;</span><span class="p">,</span>
<span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">20</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>

    <span class="c1"># UMAP</span>
    <span class="n">Subset_giotto_T</span> <span class="o">=</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">Subset_giotto_T</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">)</span>

    <span class="c1"># Leiden clustering</span>
    <span class="n">Subset_giotto_T</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">Subset_giotto_T</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>


    <span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">Subset_giotto_T</span><span class="p">,</span>
        <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s2">&quot;8b_Cluster&quot;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/8b_Cluster.png" src="../../_images/8b_Cluster.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">markers_scran_T</span> <span class="o">=</span> <span class="n">findMarkers_one_vs_all</span><span class="p">(</span><span class="n">gobject</span><span class="o">=</span><span class="n">Subset_giotto_T</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;scran&quot;</span><span class="p">,</span>
                 <span class="n">expression_values</span><span class="o">=</span><span class="s2">&quot;normalized&quot;</span><span class="p">,</span> <span class="n">cluster_column</span><span class="o">=</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">min_feats</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">markergenes_scran_T</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">markers_scran_T</span><span class="p">[,</span> <span class="n">head</span><span class="p">(</span><span class="o">.</span><span class="n">SD</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">][[</span><span class="s2">&quot;feats&quot;</span><span class="p">]])</span>

<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">Subset_giotto_T</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s2">&quot;normalized&quot;</span><span class="p">,</span> <span class="n">metadata_cols</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
                    <span class="n">selected_feats</span> <span class="o">=</span> <span class="n">markergenes_scran_T</span><span class="p">,</span>
                    <span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">show_values</span> <span class="o">=</span> <span class="s1">&#39;zscores_rescaled&#39;</span><span class="p">,</span>
                    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_c_metaheatmap&#39;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/8_c_metaheatmap.png" src="../../_images/8_c_metaheatmap.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">T_labels</span><span class="o">&lt;-</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Naive T cells&quot;</span><span class="p">,</span><span class="c1">#1</span>
    <span class="s2">&quot;Tfh cells&quot;</span><span class="p">,</span><span class="c1">#2</span>
    <span class="s2">&quot;CD8 T cells&quot;</span><span class="p">,</span><span class="c1">#3</span>
    <span class="s2">&quot;NK T cells&quot;</span><span class="p">,</span><span class="c1">#4</span>
    <span class="s2">&quot;CD4 T cells&quot;</span><span class="p">)</span><span class="c1">#5</span>
<span class="n">names</span><span class="p">(</span><span class="n">T_labels</span><span class="p">)</span><span class="o">&lt;-</span><span class="mi">1</span><span class="p">:</span><span class="mi">5</span>
<span class="n">Subset_giotto_T</span><span class="o">&lt;-</span><span class="n">annotateGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">Subset_giotto_T</span><span class="p">,</span> <span class="n">annotation_vector</span> <span class="o">=</span> <span class="n">T_labels</span> <span class="p">,</span>
                        <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;subset_labels&#39;</span><span class="p">)</span>
<span class="n">dimPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">Subset_giotto_T</span><span class="p">,</span>     <span class="n">dim_reduction_name</span> <span class="o">=</span> <span class="s1">&#39;umap&#39;</span><span class="p">,</span>
    <span class="n">cell_color</span> <span class="o">=</span> <span class="s2">&quot;subset_labels&quot;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s2">&quot;8d_Annotation&quot;</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/8d_Annotation.png" src="../../_images/8d_Annotation.png" />
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>