<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Vizgen Mouse Brain &mdash; Giotto 1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx-toolbox-code.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/sphinx-toolbox.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/tabs.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/sphinx_toolbox_installation.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script src="../../_static/sphinx_toolbox_installation.js"></script>
        <script src="../../_static/design-tabs.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html">
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contribute.html">Contribute</a></li>
<li class="toctree-l1"><a class="reference external" href="http://spatial.rc.fas.harvard.edu/giotto-viewer/">Giotto Viewer [work-in-progress]</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html">Datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../faqs.html">FAQs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../trygiotto.html">Try Giotto</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Issues</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Common Errors and Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../github_issues.html">Report a Bug</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Vizgen Mouse Brain</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/subsections/datasets/vizgen_mouse_brain.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="vizgen-mouse-brain">
<h1>Vizgen Mouse Brain<a class="headerlink" href="#vizgen-mouse-brain" title="Permalink to this headline"></a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Slide 1 Replicate 1</p>
</div>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline"></a></h2>
<p>This tutorial was written using Giotto version 2.0.0.9045. Check the version you are using to get the same results.
If you do not have Giotto installed, you can do so by running the following code.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">remotes</span><span class="p">::</span><span class="n">install_github</span><span class="p">(</span><span class="s2">&quot;RubD/Giotto@suite&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Now, load the library and define the project path.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>

<span class="c1"># 1. set working directory</span>
<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;/path/to/directory/&#39;</span>

<span class="c1"># 2. set giotto python path</span>
<span class="c1"># set python path to your preferred python version path</span>
<span class="c1"># set python path to NULL if you want to automatically install (only the 1st</span>
<span class="c1"># time) and use the giotto miniconda environment</span>
<span class="n">python_path</span> <span class="o">=</span> <span class="n">NULL</span>
<span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">python_path</span><span class="p">))</span> <span class="p">{</span>
<span class="n">installGiottoEnvironment</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</section>
<section id="dataset-explanation">
<h2>Dataset Explanation<a class="headerlink" href="#dataset-explanation" title="Permalink to this headline"></a></h2>
<p>This tutorial uses the <a class="reference external" href="https://vizgen.com/applications/neuroscience-showcase/">neuroscience showcase data</a> provided by the Vizgen company. It captures transcripts at subcellular resolution and outputs:</p>
<ul class="simple">
<li><p>List of all detected transcripts and their spatial locations in three dimensions (CSV)</p></li>
<li><p>The DAPI and Poly T mosaic images (TIFF)</p></li>
<li><p>Output from the cell segmentation analysis: the transcripts per cell matrix (CSV), the cell metadata (CSV) and the cell boundaries (HDF5)</p></li>
</ul>
<p>The company mapped the mouse brain in 9 full slices across 3 positions. However, this tutorial only uses the first replicate from the first slide.</p>
</section>
<section id="giotto-global-instructions-and-preparations">
<h2>Giotto global instructions and preparations<a class="headerlink" href="#giotto-global-instructions-and-preparations" title="Permalink to this headline"></a></h2>
<p>Define the instructions to save the plots.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># by directly saving plots, but not rendering them you will save a lot of time</span>
<span class="n">instrs</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span>
                <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                <span class="n">return_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="create-giotto-object-and-process-data">
<h2>Create Giotto object and process data<a class="headerlink" href="#create-giotto-object-and-process-data" title="Permalink to this headline"></a></h2>
<p>To create the Giotto object, you first need to read the cell by gene expression matrix and the metadata information. Since Giotto receives the expression with genes as rows and cells as columns, you need to transpose the expression matrix to create the object.</p>
<p>The Vizgen data comes with information about the field of view (FOV) and the volume of the cell. This metadata information can be added to the object.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># provide path to cell by gene matrix</span>
<span class="n">expr_path</span> <span class="o">=</span> <span class="s1">&#39;/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_cell_by_gene_S1R1.csv&#39;</span>
<span class="c1"># provide path to metadata</span>
<span class="n">spat_path</span> <span class="o">=</span> <span class="s1">&#39;/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_cell_metadata_S1R1.csv&#39;</span>

<span class="c1"># read expression matrix and metadata</span>
<span class="n">expr_matrix</span> <span class="o">=</span> <span class="n">readExprMatrix</span><span class="p">(</span><span class="n">expr_path</span><span class="p">)</span>
<span class="n">spat_dt</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="n">spat_path</span><span class="p">)</span>

<span class="c1"># create giotto object</span>
<span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">createGiottoObject</span><span class="p">(</span><span class="n">expression</span> <span class="o">=</span> <span class="n">Giotto</span><span class="p">:::</span><span class="n">t_flex</span><span class="p">(</span><span class="n">expr_matrix</span><span class="p">),</span>
                <span class="n">spatial_locs</span> <span class="o">=</span> <span class="n">spat_dt</span><span class="p">[,</span><span class="o">.</span><span class="p">(</span><span class="n">center_x</span><span class="p">,</span> <span class="n">center_y</span><span class="p">,</span> <span class="n">V1</span><span class="p">)],</span>
                <span class="n">instructions</span> <span class="o">=</span> <span class="n">instrs</span><span class="p">)</span>
<span class="c1"># add metadata of fov and volume</span>
<span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">addCellMetadata</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">new_metadata</span> <span class="o">=</span> <span class="n">spat_dt</span><span class="p">[,</span><span class="o">.</span><span class="p">(</span><span class="n">fov</span><span class="p">,</span> <span class="n">volume</span><span class="p">)])</span>
</pre></div>
</div>
<p>Visualize cells in space.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
       <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/0-spatPlot2D1.png" src="../../_images/0-spatPlot2D1.png" />
<p>Visualize cells by FOV.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;fov&#39;</span><span class="p">,</span> <span class="n">show_legend</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
       <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
       <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/1-spatPlot2D1.png" src="../../_images/1-spatPlot2D1.png" />
<p>Processing steps.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">filterGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span>
           <span class="n">expression_threshold</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
           <span class="n">feat_det_in_min_cells</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
           <span class="n">min_det_feats_per_cell</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">normalizeGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span> <span class="n">scalefactor</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
<span class="c1"># add gene and cell statistics</span>
<span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">addStatistics</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">)</span>
</pre></div>
</div>
<p>Visualize the number of features per cell.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen_brain</span><span class="p">,</span> <span class="n">show_image</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
       <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;nr_feats&#39;</span><span class="p">,</span> <span class="n">color_as_factor</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
       <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/2-spatPlot2D.png" src="../../_images/2-spatPlot2D.png" />
</section>
<section id="dimension-reduction">
<h2>Dimension Reduction<a class="headerlink" href="#dimension-reduction" title="Permalink to this headline"></a></h2>
<p>Since no HVG selection was performed, Giotto will consider all genes. The first step is to calculate the principal components.
.. code-block:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">runPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">scale_unit</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="c1"># visualize variance explained per component</span>
<span class="n">screePlot</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">ncp</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span>
</pre></div>
</div>
<img alt="../../_images/3-screePlot.png" src="../../_images/3-screePlot.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
    <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/4-PCA.png" src="../../_images/4-PCA.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vizgen</span> <span class="o">=</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
 <span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
     <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/5-UMAP.png" src="../../_images/5-UMAP.png" />
</section>
<section id="cluster">
<h2>Cluster<a class="headerlink" href="#cluster" title="Permalink to this headline"></a></h2>
<p>Calculate nearest neighbor network and perform Leiden clustering.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="n">vizgen</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<p>Visualize clusters in reduced dimension. The default cell color is <cite>‘leiden_clus’</cite>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotUMAP</span><span class="p">(</span><span class="n">vizgen</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
     <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
     <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/6-UMAP.png" src="../../_images/6-UMAP.png" />
<p>Visualize in spatial dimensions.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
       <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
       <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/7-spatPlot2D.png" src="../../_images/7-spatPlot2D.png" />
<p>It is also possible to reverse the colors for the visualization.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># get colors
cell_metadata = pDataDT(vizgen)
leiden_names = unique(cell_metadata$leiden)
leiden_colors = Giotto::getDistinctColors(n = length(leiden_names))
names(leiden_colors) = leiden_names

# reverse colors
leiden_rev_colors = Giotto::getDistinctColors(n = length(leiden_names))
names(leiden_rev_colors) = rev(leiden_names)

# visualize with reversed colors
spatPlot2D(gobject = vizgen, cell_color = &#39;leiden_clus&#39;, point_size = 0.5,
    cell_color_code = leiden_rev_colors, coord_fix_ratio = TRUE,
    background_color = &#39;black&#39;, show_plot = F, return_plot = F,
    save_plot = T, save_param = list(show_saved_plot = TRUE))
</pre></div>
</div>
<img alt="../../_images/8-spatPlot2D.png" src="../../_images/8-spatPlot2D.png" />
</section>
<section id="spatial-expression-patterns">
<h2>Spatial expression patterns<a class="headerlink" href="#spatial-expression-patterns" title="Permalink to this headline"></a></h2>
<p>The first step is to calculate the spatial network and then perform the binary spatial extraction of genes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># create spatial network based on physical distance of cell centroids
vizgen = createSpatialNetwork(gobject = vizgen, minimum_k = 2,
                maximum_distance_delaunay = 50)

# select features
feats = vizgen@feat_ID$rna
# perform Binary Spatial Extraction of genes
km_spatialgenes = binSpect(vizgen, subset_feats = feats)

# visualize spatial expression of selected genes obtained from binSpect
spatFeatPlot2D(vizgen, expression_values = &#39;scaled&#39;,
        feats = c(&#39;Slc47a1&#39;, &#39;Slc17a7&#39;, &#39;Th&#39;, &#39;Npy2r&#39;, &#39;Chrm1&#39;, &#39;Gfap&#39;),
        cell_color_gradient = c(&#39;blue&#39;, &#39;white&#39;, &#39;red&#39;),
        point_shape = &#39;border&#39;, point_border_stroke = 0.01,
        show_network = F, network_color = &#39;lightgrey&#39;, point_size = 0.2,
        cow_n_col = 2)
</pre></div>
</div>
<img alt="../../_images/9-spatFeatPlot2D.png" src="../../_images/9-spatFeatPlot2D.png" />
</section>
<section id="subset-giotto-and-add-cell-boundary-information">
<h2>Subset Giotto and add cell boundary information<a class="headerlink" href="#subset-giotto-and-add-cell-boundary-information" title="Permalink to this headline"></a></h2>
<p>Giotto can be subset to analyze only a portion of the data.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">vizgen_subset</span> <span class="o">&lt;-</span> <span class="n">subsetGiottoLocs</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen</span><span class="p">,</span>
                  <span class="n">x_min</span> <span class="o">=</span> <span class="mi">2000</span><span class="p">,</span> <span class="n">x_max</span> <span class="o">=</span> <span class="mi">3000</span><span class="p">,</span>
                  <span class="n">y_max</span> <span class="o">=</span> <span class="mi">3500</span><span class="p">,</span> <span class="n">y_min</span> <span class="o">=</span> <span class="mi">2500</span><span class="p">)</span>
</pre></div>
</div>
<p>The visualization functions can also be applied to the subset version.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen_subset</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">show_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">save_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
    <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/10-spatPlot2D.png" src="../../_images/10-spatPlot2D.png" />
<p>Giotto can include the information about the polygons as provided by Vizgen. Since we are working with a subset of the data, it is necessary to read only the polygons that are present in the current FOVs.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># define path to cell boundaries folder</span>
<span class="n">bound_path</span> <span class="o">=</span> <span class="s1">&#39;/path/to/cell_boundaries&#39;</span>
<span class="c1"># read polygons and add them to Giotto</span>
<span class="n">vizgen_subset</span> <span class="o">=</span> <span class="n">readPolygonFilesVizgen</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">vizgen_subset</span><span class="p">,</span>
                    <span class="n">boundaries_path</span> <span class="o">=</span> <span class="n">bound_path</span><span class="p">,</span>
                    <span class="n">polygon_feat_types</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
</pre></div>
</div>
<p>Giotto can also include information about the transcripts.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># add transcript coordinates
tx_path = &#39;/path/to/datasets_mouse_brain_map_BrainReceptorShowcase_Slice1_Replicate1_detected_transcripts_S1R1.csv&#39;
tx_dt = data.table::fread(tx_path)
# select transcripts in FOVs
selected_fovs = unique(pDataDT(vizgen_subset)$fov)
tx_dt_selected = tx_dt[fov %in% selected_fovs]

# create Giotto points from transcripts
gpoints = createGiottoPoints(x = tx_dt_selected[,.(global_x,global_y, gene)])

# add points to Giotto
vizgen_subset = addGiottoPoints(gobject = vizgen_subset,
                gpoints = list(gpoints))

# identify genes for visualization
gene_meta = fDataDT(vizgen_subset)
data.table::setorder(gene_meta, perc_cells)
gene_meta[perc_cells &gt; 25 &amp; perc_cells &lt; 50]

# visualize points from index z0
spatInSituPlotPoints(vizgen_subset,
            feats = list(&#39;rna&#39; = c(&quot;Oxgr1&quot;, &quot;Htr1a&quot;, &quot;Gjc3&quot;, &quot;Axl&quot;,
                        &#39;Gfap&#39;, &quot;Olig1&quot;, &quot;Epha7&quot;)),
            polygon_feat_type = &#39;z0&#39;,
            use_overlap = F,
            point_size = 0.2,
            show_polygon = TRUE,
            polygon_color = &#39;white&#39;,
            return_plot = FALSE,
            save_plot = TRUE,
            show_plot = FALSE,
            save_param = list(show_saved_plot = TRUE))
</pre></div>
</div>
<img alt="../../_images/11-spatInSituPlotPoints.png" src="../../_images/11-spatInSituPlotPoints.png" />
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># visualize points from index z6</span>
<span class="n">spatInSituPlotPoints</span><span class="p">(</span><span class="n">vizgen_subset</span><span class="p">,</span>
            <span class="n">feats</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s1">&#39;rna&#39;</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Oxgr1&quot;</span><span class="p">,</span> <span class="s2">&quot;Htr1a&quot;</span><span class="p">,</span> <span class="s2">&quot;Gjc3&quot;</span><span class="p">,</span> <span class="s2">&quot;Axl&quot;</span><span class="p">,</span>
                        <span class="s1">&#39;Gfap&#39;</span><span class="p">,</span> <span class="s2">&quot;Olig1&quot;</span><span class="p">,</span> <span class="s2">&quot;Epha7&quot;</span><span class="p">)),</span>
            <span class="n">polygon_feat_type</span> <span class="o">=</span> <span class="s1">&#39;z6&#39;</span><span class="p">,</span>
            <span class="n">use_overlap</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span>
            <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
            <span class="n">show_polygon</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
            <span class="n">polygon_color</span> <span class="o">=</span> <span class="s1">&#39;white&#39;</span><span class="p">,</span>
            <span class="n">return_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
            <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
            <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">show_saved_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<img alt="../../_images/12-spatInSituPlotPoints1.png" src="../../_images/12-spatInSituPlotPoints1.png" />
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>